sdk/eventhubs/microsoft-azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/impl/MessagingFactory.java

```diff
@@ -248,6 +283,100 @@ public void run() {
+            TRACE_LOGGER.info("Watchdog scheduling first run");
+            this.watchdogFuture = this.executor.schedule(new WatchDog(), this.watchdogScanSeconds, TimeUnit.SECONDS);
+        } else {
+            TRACE_LOGGER.info("Watchdog is OFF");
+        }
+    }
+
+    private class WatchDog implements Runnable {
+        @Override
+        public void run() {
+            TRACE_LOGGER.debug("Watchdog run");
+            if (MessagingFactory.this.getIsClosingOrClosed()) {
+                return;
+            }
+            if (MessagingFactory.this.watchdogTriggerSeconds <= EventHubClientOptions.SILENT_OFF) {
+                TRACE_LOGGER.warn("Watchdog should not run when trigger time is " + MessagingFactory.this.watchdogTriggerSeconds + " -- stopping");
+                return;
+            }
+
+            LinkedList<MessageReceiver> copiedList = null;
+            synchronized (MessagingFactory.this.watchdogSyncObject) {
+                copiedList = new LinkedList<MessageReceiver>(MessagingFactory.this.watchdogReceivers);
+            }
+            if (!copiedList.isEmpty()) {
+                boolean anyReceiverIsAlive = false;
+                final long longestAgoAllowable = Instant.now().getEpochSecond()
+                        - MessagingFactory.this.watchdogTriggerSeconds;
+
+                for (MessageReceiver rcvr : copiedList) {
+                    TRACE_LOGGER.debug("Watchdog checking receiver " + rcvr.getClientId() + " last: "
+                            + rcvr.getLastReceivedTime() + "  allowable: " + longestAgoAllowable);
+                    if (!rcvr.getIsClosingOrClosed() && (rcvr.getLastReceivedTime() >= longestAgoAllowable)) {
+                        anyReceiverIsAlive = true;
+                        // Found one live receiver, no need to check the rest.
+                        break;
+                    }
+                }
+
+                if (!anyReceiverIsAlive && !MessagingFactory.this.getIsClosingOrClosed()) {
+                    TRACE_LOGGER.warn("Watchdog forcing connection closed");
+                    ErrorCondition suspect = new ErrorCondition(ClientConstants.WATCHDOG_ERROR,
+                            "receiver watchdog has fired, all receivers silent");
+                    MessagingFactory.this.watchdogCleanupDone = false;
+                    MessagingFactory.this.connection.setCondition(suspect);
+                    MessagingFactory.this.connection.close();
+                    // If the remote host is still responding at the TCP level, then the socket will
+                    // close normally and cleanup/recreation will happen automatically. However, if it
+                    // isn't, then we must call onConnectionError here in order to force cleanup and
+                    // recreation.
+                    try {
+                        Thread.sleep(5000);
+                    } catch (InterruptedException e) {
+                    }
```

Comment:
What happens when the executor is single-threaded? Having a Thread.sleep() on this thread can stall other operations scheduled in this executor.
`executor = Executors.newScheduledThreadPool(1);`

Reply:
A single-threaded executor will not perform adequately for the scenarios that this watchdog is intended for. Anyone who should be using this feature will have a real thread pool. If anyone else is fooling around with the watchdog without consulting us, the worst that can happen is five-second pauses.